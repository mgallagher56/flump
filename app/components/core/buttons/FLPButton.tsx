import React from 'react';

import type { ButtonProps } from '@chakra-ui/react';
import { Button } from '@chakra-ui/react';
import * as pressable from '@zag-js/pressable';
import type { PressEvent } from '@zag-js/pressable/dist/pressable.types';
import { normalizeProps, useMachine } from '@zag-js/react';

interface FLPButtonProps extends ButtonProps {
  preventFocusOnPress?: boolean;
  onPress?: (e: PressEvent) => void;
  onLongPress?: (e: PressEvent) => void;
}

const FLPButton: React.FC<React.PropsWithChildren<FLPButtonProps>> = ({
  children,
  colorScheme = 'blue',
  isDisabled,
  preventFocusOnPress,
  variant = 'solid',
  onLongPress,
  onPress,
  isLoading
}) => {
  const [state, send] = useMachine(
    pressable.machine({
      id: 'pressableBaseButton',
      onPress,
      onLongPress,
      disabled: isDisabled,
      preventFocusOnPress
    })
  );

  const api = pressable.connect(state, send, normalizeProps);

  return (
    <Button {...api.pressableProps} variant={variant} colorScheme={colorScheme} isLoading={isLoading}>
      {children}
    </Button>
  );
};

export default FLPButton;
