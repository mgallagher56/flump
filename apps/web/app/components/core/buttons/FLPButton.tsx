import { type FC, type PropsWithChildren } from 'react';

import type { ButtonProps } from '@chakra-ui/react';
import { Button } from '@chakra-ui/react';

interface FLPButtonProps extends ButtonProps {
  padding?: number;
}

const FLPButton: FC<PropsWithChildren<FLPButtonProps>> = ({
  colorScheme = 'blue',
  disabled,
  isLoading,
  padding,
  variant = 'solid',
  onClick,
  ...buttonProps
}) => {
  return (
    <Button
      colorScheme={colorScheme}
      isDisabled={disabled}
      isLoading={isLoading}
      padding={padding}
      variant={variant}
      onClick={onClick}
      width={'max-content'}
      {...buttonProps}
    >
      {buttonProps.children}
    </Button>
  );
};

export default FLPButton;
