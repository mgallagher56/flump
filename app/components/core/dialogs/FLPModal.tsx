import { cloneElement, useCallback, useId } from 'react';
import type { FC, ReactElement } from 'react';

import {
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay
} from '@chakra-ui/react';
import * as dialog from '@zag-js/dialog';
import { mergeProps, normalizeProps, useMachine } from '@zag-js/react';
import { useTranslation } from 'react-i18next';

import FLPButton from '../buttons/FLPButton';
import FLPButtonGroup from '../buttons/FLPButtonGroup';
import FLPBox from '../structure/FLPBox';

interface FLPModalProps {
  triggerBtn: ReactElement;
  confirmButton?: {
    text: string;
    colorScheme?: string;
    variant?: string;
  };
  content: ReactElement;
  isDisabled?: boolean;
  title: string;
  onConfirm: (e) => Promise<void>;
}

const FLPModal: FC<FLPModalProps> = ({
  triggerBtn,
  confirmButton,
  content,
  isDisabled,
  title,
  onConfirm,
  ...props
}) => {
  const { t } = useTranslation();
  const [state, send] = useMachine(dialog.machine({ id: useId() }));

  const api = dialog.connect(state, send, normalizeProps);
  const mergedProps = mergeProps({
    ...api,
    ...props,
    backdropProps: {
      ...api.backdropProps,
      onDrag: undefined,
      onDragEnd: undefined,
      onDragStart: undefined,
      onAnimationStart: undefined
    }
  });

  const handleOnConfirm = useCallback(
    (event) => {
      onConfirm(event);
      send('CONFIRM');
      mergedProps.closeTriggerProps.onClick(event);
    },
    [mergedProps.closeTriggerProps, onConfirm, send]
  );

  const triggerButtonClone = cloneElement(triggerBtn, {
    ...api.triggerProps
  });

  return (
    <>
      {triggerButtonClone}
      {mergedProps.isOpen && (
        <Modal isOpen={mergedProps.isOpen} onClose={mergedProps.close} isCentered>
          <ModalOverlay
            {...mergedProps.backdropProps}
            onDrag={undefined}
            onDragEnd={undefined}
            onDragStart={undefined}
            onAnimationStart={undefined}
          />
          <FLPBox {...mergedProps.containerProps}>
            <ModalContent {...mergedProps.contentProps}>
              <ModalHeader {...mergedProps.titleProps}>{title}</ModalHeader>
              <ModalCloseButton {...mergedProps.closeTriggerProps} />
              <ModalBody {...mergedProps.descriptionProps}>{content}</ModalBody>
              <ModalFooter>
                <FLPButtonGroup>
                  <FLPButton variant="outline" {...mergedProps.closeTriggerProps}>
                    {t('close')}
                  </FLPButton>
                  {onConfirm && (
                    <FLPButton
                      colorScheme={confirmButton?.colorScheme}
                      isDisabled={isDisabled}
                      variant={confirmButton?.variant}
                      onClick={handleOnConfirm}
                    >
                      {confirmButton.text}
                    </FLPButton>
                  )}
                </FLPButtonGroup>
              </ModalFooter>
            </ModalContent>
          </FLPBox>
        </Modal>
      )}
    </>
  );
};

export default FLPModal;
