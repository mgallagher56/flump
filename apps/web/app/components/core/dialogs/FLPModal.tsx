import { type FC, type ReactElement, useCallback } from 'react';

import { type DialogRootProps } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';
import FLPButton from '~/components/core/buttons/FLPButton';
import FLPButtonGroup from '~/components/core/buttons/FLPButtonGroup';
import {
  DialogBackdrop,
  DialogBody,
  DialogCloseTrigger,
  DialogContent,
  DialogFooter,
  DialogHeader,
  DialogRoot,
  DialogTitle,
  DialogTrigger
} from '~/components/ui/dialog';

interface FLPModalProps extends DialogRootProps {
  triggerBtn: ReactElement;
  confirmButton?: {
    text: string;
    colorPalette?: string;
    id?: string;
  };
  contentRef?: React.RefObject<HTMLDivElement>;
  disabled?: boolean;
  open: boolean;
  title: string;
  onConfirm: () => Promise<void>;
  onClose: () => void;
}

const FLPModal: FC<FLPModalProps> = ({
  children,
  confirmButton,
  contentRef,
  disabled,
  open,
  title,
  triggerBtn,
  onClose,
  onConfirm
}) => {
  const { t } = useTranslation();

  const handleOnConfirm = useCallback(() => {
    onConfirm();
    onClose();
  }, [onClose, onConfirm]);

  return (
    <DialogRoot
      closeOnEscape
      unmountOnExit
      motionPreset="slide-in-bottom"
      open={open}
      placement="center"
      onOpenChange={onClose}
    >
      <DialogBackdrop />
      <DialogTrigger>{triggerBtn}</DialogTrigger>
      <DialogContent ref={contentRef}>
        <DialogHeader>
          <DialogTitle>{title}</DialogTitle>
        </DialogHeader>
        <DialogBody>{children}</DialogBody>
        <DialogFooter>
          <FLPButtonGroup>
            <FLPButton variant="outline" onClick={onClose}>
              {t('close')}
            </FLPButton>
            {onConfirm && (
              <FLPButton
                colorPalette={confirmButton?.colorPalette}
                disabled={disabled}
                form={confirmButton?.id}
                type="submit"
                onClick={handleOnConfirm}
              >
                {confirmButton.text}
              </FLPButton>
            )}
          </FLPButtonGroup>
          <DialogCloseTrigger />
        </DialogFooter>
      </DialogContent>
    </DialogRoot>
  );
};

export default FLPModal;
