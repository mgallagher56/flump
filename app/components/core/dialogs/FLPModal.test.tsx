import { fireEvent, render } from '@testing-library/react';
import { vi } from 'vitest';

import FLPButton from '../buttons/FLPButton';
import FLPModal from './FLPModal';

const mocks = vi.hoisted(() => ({
  mockOnConfirm: vi.fn()
}));

describe('<FLPModal />', () => {
  const { baseElement, getByText } = render(
    <FLPModal
      children={<div>Modal Content</div>}
      confirmButton={{
        text: 'Confirm',
        colorScheme: 'blue',
        variant: 'solid'
      }}
      title="Modal Title"
      triggerBtn={<FLPButton>Trigger</FLPButton>}
      onConfirm={mocks.mockOnConfirm}
    />
  );
  const triggerBtn = getByText('Trigger');
  fireEvent.click(triggerBtn);
  it('renders by default as expected', () => {
    const confirmButton = getByText('Confirm');
    fireEvent.click(confirmButton);
    expect(mocks.mockOnConfirm).toHaveBeenCalled();
    const htmlString = baseElement.outerHTML.toString();
    const baseElementConstant = htmlString.replaceAll(/style="[^"]*"/g, '');
    expect(baseElementConstant).toMatchSnapshot();
  });
});
