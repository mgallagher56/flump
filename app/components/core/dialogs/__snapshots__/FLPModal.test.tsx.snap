// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<FLPModal /> > renders by default as expected 1`] = `"<body><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><button type=\\"button\\" class=\\"chakra-button css-1ltlu47\\">Trigger</button></div><div class=\\"chakra-portal\\"><div class=\\"chakra-modal__overlay css-13d1tkw\\"  data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"></div><div data-focus-guard=\\"true\\" tabindex=\\"0\\"  data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"></div><div data-focus-lock-disabled=\\"false\\"><div class=\\"chakra-modal__content-container css-1oxhx2p\\" tabindex=\\"-1\\"><section class=\\"chakra-modal__content css-6yiygt\\" role=\\"dialog\\" id=\\"chakra-modal-:r0:\\" tabindex=\\"-1\\" aria-modal=\\"true\\"  aria-labelledby=\\"chakra-modal--header-:r0:\\" aria-describedby=\\"chakra-modal--body-:r0:\\"><header class=\\"chakra-modal__header css-72fd9l\\" id=\\"chakra-modal--header-:r0:\\">Modal Title</header><button type=\\"button\\" aria-label=\\"Close\\" class=\\"chakra-modal__close-btn css-6sdaxy\\"><svg viewBox=\\"0 0 24 24\\" focusable=\\"false\\" class=\\"chakra-icon css-onkibi\\" aria-hidden=\\"true\\"><path fill=\\"currentColor\\" d=\\"M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z\\"></path></svg></button><div class=\\"chakra-modal__body css-0\\" id=\\"chakra-modal--body-:r0:\\"><div>Modal Content</div></div><footer class=\\"chakra-modal__footer css-192qrng\\"><div role=\\"group\\" class=\\"chakra-button__group css-1ind3v2\\" data-orientation=\\"horizontal\\"><button type=\\"button\\" class=\\"chakra-button css-kn8t1u\\">close</button><button type=\\"submit\\" class=\\"chakra-button css-kn8t1u\\">Confirm</button></div></footer></section></div></div><div data-focus-guard=\\"true\\" tabindex=\\"0\\"  data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"></div></div></body>"`;
